Deprecate Time module and all its Schedulers

EwoutH opened the pull request:
As concluded in https://github.com/projectmesa/mesa/discussions/2231, this PR deprecates the whole Time module and all its Schedulers. This change aims to provide more flexibility and control in agent activation and time management, inspired by the NetLogo framework.

### Changes
1. Deprecated the entire `mesa.time` module.
2. Added a module-level deprecation warning to `mesa.time`.
3. Updated the migration guide to include instructions for replacing each scheduler with AgentSet functionality.
4. Removed the `tests/test_time.py` file as it's no longer relevant.

### Migration Guide
The migration guide has been updated to include detailed instructions on how to replace each scheduler with AgentSet functionality. This includes examples for:

- BaseScheduler
- RandomActivation
- SimultaneousActivation
- StagedActivation
- RandomActivationByType

The guide also includes general notes on changes to related functionality, such as the automatic incrementing of `Model.steps` and the replacement of scheduler-specific methods with AgentSet methods.

For this PR, a rendered version of the migration guide can be [viewed here](https://mesa--2306.org.readthedocs.build/en/2306/migration_guide.html#time-and-schedulers).

### Benefits
As discussed extensively in [#1912](https://github.com/projectmesa/mesa/discussions/1912) and [#2231](https://github.com/projectmesa/mesa/discussions/2231), replacing the time module with AgentSet functionality has several benefits. A short summary:

1. **Flexibility**: Users can now create custom activation patterns without being restricted to predefined schedulers.
2. **Explicitness**: The behavior of agent activation is now explicitly defined in the model, improving readability and maintainability.
3. **Simplicity**: The codebase is simplified by removing the need for multiple scheduler classes.
4. **Consistency**: This change aligns Mesa more closely with other agent-based modeling frameworks like NetLogo.

### References
- Original discussion: [#1912](https://github.com/projectmesa/mesa/discussions/1912)
- Decision discussion: [#2231](https://github.com/projectmesa/mesa/discussions/2231)
- Example updates: [#183](https://github.com/projectmesa/mesa-examples/pull/183) and [#201](https://github.com/projectmesa/mesa-examples/pull/201)
- This PR: [#2306](https://github.com/projectmesa/mesa/pull/2306)
- Faster alternative for `.shuffle().do()`: [#2283](https://github.com/projectmesa/mesa/pull/2283)

### PR checklist
- [x] Write migration guide
- [x] Update this PR message
- [x] Find a faster alternative for `.shuffle().do()` (see [#2283](https://github.com/projectmesa/mesa/pull/2283))

### Next Steps
- Update all example models to use the new AgentSet functionality instead of schedulers.
- Review and update any remaining documentation that references the deprecated schedulers.

rht commented on mesa/time.py:
```diff
@@ -31,6 +31,15 @@
 
+# Throw module-level deprecation warning
+warnings.warn(
+    "The time module and all its Schedulers are deprecated and will be removed in a future version. "
```
Does this mean 3.1 or 4.0?

quaquel replied:
I would not put a timeline on its removal. 

Just deprecation is enough. We don't need to worry about maintaining and testing anymore, but it might be convenient to keep it around for a while to make it easier for people to migrate models.

EwoutH replied:
Yeah good question, I'm not sure about that. I'm nudging towards 3.1 or 3.2.

Another option is leaving it as a warning for now, and flipping it to an Error at some point.

EwoutH replied:
I wrote the comment before I read Jan's comment, but I agree with him. We don't support it anymore and it will get removed.

EwoutH commented:
Updated the migration guide, so this PR is ready for review.

A rendered version of the migration guide can be [viewed here](https://mesa--2306.org.readthedocs.build/en/2306/migration_guide.html#time-and-schedulers).

rht commented on mesa/time.py:
```diff
@@ -31,6 +31,15 @@
 from mesa.model import Model
 
+# Throw module-level deprecation warning
+warnings.warn(
```
Since time.py is imported into `__init__.py`, this will throw a warning all the time `mesa` is imported. This will be too noisy.

quaquel replied:
That's a good catch, and I agree. the fix is to remove it from the overall init then?

rht replied:
One option is to add the warning to every classes `__init__`.

EwoutH replied:
Would the warning in the Basescheduler work for now? Since all other schedulers use it.

EwoutH replied:
Moved.

EwoutH replied:
Doesn’t really seem to solve it.

quaquel replied:
It should go into the init itself. A class is parsed on import so currently the error is thrown at that moment already. In the init the error gets thrown when initializing the class.

EwoutH replied:
Done, and I checked that all other schedulers call the Baseclass init, which is the case.

EwoutH commented:
Moved the DepreciationWarning, and also added a nice visible warning to the docstring.

![IMG_1466](https://github.com/user-attachments/assets/f2c214e5-3671-4f42-8d64-c71cfd3c56be)

Merging, will work on updating the examples next.
